### PR TITLE
Performance tuning documentation

### PR DESCRIPTION
**Pull request by zimnx**

### CODE REVIEW

File: docs/source/performance.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+   nodeSelector:
+     scylla.scylladb.com/node-type: scylla
+```
+For more details about new CRD use:
+```
+kubectl explain nodeconfigs.scylla.scylladb.com/v1alpha1
+```
+
+Tuning consists of two separate optimizations: common tuning of node, and tuning based on Scylla Pods and their resource assignment.
+First optimization is executed when tuning Pod lands on Node. Second one is executed when Scylla Pod lands on the same Node.
```
Comment (tnozicka) [inline docs/source/performance.md:27]: not a fan of those long references (first, second) - would it be better to replace them with "node tuning" and "pod tuining?

File: docs/source/performance.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+     scylla.scylladb.com/node-type: scylla
+```
+For more details about new CRD use:
+```
+kubectl explain nodeconfigs.scylla.scylladb.com/v1alpha1
+```
+
+Tuning consists of two separate optimizations: common tuning of node, and tuning based on Scylla Pods and their resource assignment.
+First optimization is executed when tuning Pod lands on Node. Second one is executed when Scylla Pod lands on the same Node.  
+Tuning Pods are created in special namespace called `scylla-operator-node-tuning`
```
Comment (tnozicka) [inline docs/source/performance.md:28]:
```suggestion
Tuning resources are created in special namespace called `scylla-operator-node-tuning`
```

File: docs/source/performance.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+Only Scylla Pods having Guaranteed Quality of Service class are tuned, so make sure your ScyllaCluster meets all necessary conditions.
```
Comment (tnozicka) [inline docs/source/performance.md:30]: The tuning is applied only to pods with *guaranted* QoS. Please double check you ScyllaCluster resource specifications.

File: docs/source/performance.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+## Kubernetes tuning
+
+By default, the kubelet uses the CFS quota to enforce pod CPU limits.  
+When the node runs many CPU-bound pods, the workload can move around different CPU cores depending on whether the pod 
+is throttled and which CPU cores are available.
+However, kubelet may be configured to assign CPUs exclusively, by setting the CPU manager policy to static.
+
+Consult your Kubernetes provider how to set it up, because it's configurable on kubelet startup.
```
Comment (tnozicka) [inline docs/source/performance.md:39]:
```suggestion
Setting up kubelet configuration is provider specific. Please check the docs for your distribution or talk to your provider.
```

File: docs/source/performance.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    - name: us-east-1a
+      members: 3
+      storage:
+        capacity: 500Gi
+      agentResources:
+        limits:
+          cpu: 1
+          memory: 1G
+      resources:
+        limits:
```
Comment (tnozicka) [inline docs/source/performance.md:72]: maybe it would be good to set request explicitly so it's seen they are the same

Comment (zimnx) [reply docs/source/performance.md:80]: done

File: docs/source/performance.md (hunk 4 above)
Comment (zimnx) [reply docs/source/performance.md]: done

File: docs/source/performance.md (hunk 3 above)
Comment (zimnx) [reply docs/source/performance.md]: done

File: docs/source/performance.md (hunk 2 above)
Comment (zimnx) [reply docs/source/performance.md]: done

File: docs/source/performance.md (hunk 1 above)
Comment (zimnx) [reply docs/source/performance.md]: done

File: docs/source/performance.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
+# Performance tuning
+
+1.6 Scylla Operator adds a new experimental feature allowing users to optimize Kubernetes nodes
```
[the remark this answers is not in the document]
Comment (zimnx) [reply docs/source/performance.md]: done

File: docs/source/performance.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
+# Performance tuning
+
+Scylla Operator 1.6 introduces a new experimental feature allowing users to optimize Kubernetes nodes.
+
+## Node tuning
+
+From 1.6 a new CRD called NodeConfig is available which allows users to target Nodes which should be tuned.
```
Comment (tzach) [inline docs/source/performance.md:7]:
```suggestion
Starting from Operator 1.6, a new CRD called NodeConfig is available, allowing users to target Nodes that should be tuned.
```

File: docs/source/performance.md [hunk 8]
```diff
@@ -0,0 +1,83 @@
+# Performance tuning
+
+Scylla Operator 1.6 introduces a new experimental feature allowing users to optimize Kubernetes nodes.
+
+## Node tuning
+
+From 1.6 a new CRD called NodeConfig is available which allows users to target Nodes which should be tuned.  
+Nodes matching the provided placement conditions will be subject for tuning.
```
Comment (tzach) [inline docs/source/performance.md:8]:
```suggestion
Nodes matching the provided placement conditions will be subject to tuning.
```

File: docs/source/performance.md [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+ placement:
+   nodeSelector:
+     scylla.scylladb.com/node-type: scylla
+```
+For more details about new CRD use:
+```
+kubectl explain nodeconfigs.scylla.scylladb.com/v1alpha1
+```
+
+Tuning consists of two separate optimizations: common tuning of node, and tuning based on Scylla Pods and their resource assignment.
```
Comment (tzach) [inline docs/source/performance.md:26]:
```suggestion
Tuning consists of two separate optimizations: common node tuning, and tuning based on Scylla Pods and their resource assignment.
```

File: docs/source/performance.md [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+Node tuning is executed when tuning Pod lands on Node. Pod tuning is executed when Scylla Pod lands on the same Node.  
+Tuning resources are created in special namespace called `scylla-operator-node-tuning`
```
Comment (tzach) [inline docs/source/performance.md:28]:
```suggestion
Tuning resources are created in a special namespace called `scylla-operator-node-tuning`
```

File: docs/source/performance.md [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+However, kubelet may be configured to assign CPUs exclusively, by setting the CPU manager policy to static.
+
+Setting up kubelet configuration is provider specific. Please check the docs for your distribution or talk to your
+provider.
+
+Pods which receive a Guaranteed Quality of Service class can take advantage of this option. 
+As Guaranteed pods whose containers fit the requirements for being statically assigned are scheduled to the node, 
+CPUs are removed from the shared pool and placed in the cpuset for the container.
+
+In our case there are two requirements each ScyllaCluster must fulfill in order to receive a Guaranteed QoS class:
```
Comment (tzach) [inline docs/source/performance.md:47]:
```suggestion
In our case there are two requirements each ScyllaCluster must fulfill to to receive a Guaranteed QoS class:
```

File: docs/source/performance.md [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+* resource request and limits must be equal or only limits have to be provided
+* agentResources must be provided and their requests and limits must be equal, or only limits have to be provided
+
+An example of such a ScyllaCluster which receives a Guaranteed QoS class is below:
```
Comment (tzach) [inline docs/source/performance.md:51]:
```suggestion
An example of such a ScyllaCluster that receives a Guaranteed QoS class is below:
```

**Review by tzach — COMMENTED** [review]
Subset of Grammarly suggestion

**Comment by tzach** [comment]
Doc explains how to set and use Performance tuning, but give nothing on what tuning are done, why why they help

File: docs/source/performance.md (hunk 7 above)
Comment (zimnx) [reply docs/source/performance.md]: done

File: docs/source/performance.md (hunk 8 above)
Comment (zimnx) [reply docs/source/performance.md]: done

File: docs/source/performance.md (hunk 10 above)
Comment (zimnx) [reply docs/source/performance.md]: done

File: docs/source/performance.md (hunk 11 above)
Comment (zimnx) [reply docs/source/performance.md]: done

File: docs/source/performance.md (hunk 12 above)
Comment (zimnx) [reply docs/source/performance.md]: done

**Comment by zimnx** [comment]
> Doc explains how to set and use Performance tuning, but give nothing on what tuning are done, why why they help

I added mention about which optimizations perftune has, and what is the main source of interrupts. 
There's also a line about that exclusively assigned CPUs are removed from shared CPU pool.

**Comment by tzach** [comment]
LGTM

File: docs/source/performance.md [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+   nodeSelector:
+     scylla.scylladb.com/node-type: scylla
+```
+For more details about new CRD use:
+```
+kubectl explain nodeconfigs.scylla.scylladb.com/v1alpha1
+```
+
+For all optimizations we use a Python script available in the Scylla image called perftune. 
+Perftune - contains performance optimization - it’s able to tune the kernel, network and disk devices, spread IRQs across CPUs and more.
```
Comment (tnozicka) [inline docs/source/performance.md:28]:
```suggestion
Perftune executes the performance optimizations like tuning the kernel, network and disk devices, spreading IRQs across CPUs and more.
```

File: docs/source/performance.md [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+Tuning consists of two separate optimizations: common node tuning, and tuning based on Scylla Pods and their resource assignment.
+Node tuning is executed when tuning Pod lands on Node. Pod tuning is executed when Scylla Pod lands on the same Node.
```
Comment (tnozicka) [inline docs/source/performance.md:31]:
```suggestion
Node tuning is executed immediately. Pod tuning is executed when Scylla Pod lands on the same Node.
```

File: docs/source/performance.md [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+Scylla does not like sharing, it works most efficiently when it’s not interrupted.
```
Comment (tnozicka) [inline docs/source/performance.md:33]:
```suggestion
Scylla works most efficiently when it's pinned to a CPU and not interrupted. 
```

File: docs/source/performance.md [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+
+By default, the kubelet uses the CFS quota to enforce pod CPU limits.  
+When the node runs many CPU-bound pods, the workload can move around different CPU cores depending on whether the pod 
+is throttled and which CPU cores are available.
+However, kubelet may be configured to assign CPUs exclusively, by setting the CPU manager policy to static.
+
+Setting up kubelet configuration is provider specific. Please check the docs for your distribution or talk to your
+provider.
+
+Pods which receive a Guaranteed Quality of Service class can take advantage of this option.
```
Comment (tnozicka) [inline docs/source/performance.md:56]:
```suggestion
Only pods within the [Guaranteed QoS class](https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/#create-a-pod-that-gets-assigned-a-qos-class-of-guaranteed) can take advantage of this option. 
```

File: docs/source/performance.md [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+When the node runs many CPU-bound pods, the workload can move around different CPU cores depending on whether the pod 
+is throttled and which CPU cores are available.
+However, kubelet may be configured to assign CPUs exclusively, by setting the CPU manager policy to static.
+
+Setting up kubelet configuration is provider specific. Please check the docs for your distribution or talk to your
+provider.
+
+Pods which receive a Guaranteed Quality of Service class can take advantage of this option. 
+As Guaranteed pods whose containers fit the requirements for being statically assigned are scheduled to the node, 
+CPUs are removed from the shared pool and placed in the cpuset for the container.
```
Comment (tnozicka) [inline docs/source/performance.md:58]:
```suggestion
When such pods land on a Node, kubelet will pin them to specific CPUs.
```

File: docs/source/performance.md [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+
+In our case there are two requirements each ScyllaCluster must fulfill to receive a Guaranteed QoS class:
```
Comment (tnozicka) [inline docs/source/performance.md:60]: would it be better to adjust the working to match https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/#create-a-pod-that-gets-assigned-a-qos-class-of-guaranteed ?

**Review by tnozicka — APPROVED** [review]

File: docs/source/performance.md (hunk 18 above)
Comment (zimnx) [reply docs/source/performance.md:60]: I would like to leave explicit mention about agentResources, because users are not aware of it

**Pull request merged**